README
TalkSoup is a fully-featured IRC client written with the GNUstep framework

===============================================================================
Copyright (C) 2003 Andy Ruder

  TalkSoup is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Comments are welcome.
===============================================================================

****************************************************************
* netclasses MUST be installed before installation of TalkSoup *
****************************************************************

**** Please see README.osx for instructions dealing with OS X compilation ****

============================
Compilation/Installation/Use
============================

1. Compile and install TalkSoup
 andy ~/TalkSoup $ make
 andy ~/TalkSoup $ su -c 'make install'

2. Run TalkSoup
 andy ~ $ gopen TalkSoup.app 
  
Enjoy!  Read below for more information about what you've just installed
and are running :)

===========
Description
===========
TalkSoup.app is a fully-featured IRC client which uses GNUstep base.  While
TalkSoup.app itself is just a very basic core which can simply pass messages
between bundles (plugins), there are many bundles which make up the
functionality.

This distribution comes with:

Input Plugins:

  *NetclassesInput: an input plugin based on netclasses

Output Plugins:

  *GNUstepOutput: an output plugin based on gnustep-gui

Input Filters:
  
  *Highlighting: highlights keywords and nicknames and 
    handles highlighting tabs as well.  See the /highlighting
    command for configuration options. (Bundle must be loaded)

  *KeepAlive: sends a ping to the server every three minutes
    on every connection to keep you connected
  
  *DCCSupport: allows files to be sent and received from client
    to client.  There is no DCC chat support and probably never
    will be.  See /dcc command once the bundle is loaded for
    more information.

  Logging: allows a log of a connection to be made.  See the
    /logging command for more information (once the bundle
    is loaded).  Also see FAQ question 3.5 on using the logging
    support.
  
  MessageInTab: opens a new tab when someone private
    messages you

  Decolorizer: removes colors from incoming messages.

  Debold: removes bold from incoming messages

  Deunderline: removes underlining from incoming messages
  
Output Filters:

  *Colorizer: Lets you use %B for bold, %U for underline,
    %R for reverse, %O for clear all attributes, and 
    %C(0-15),(0-15) to set colors. (First number is the 
    foreground color, second number is the background color).

  Piper: Lets you pipe outgoing messages through arbitrary
    commands. See the /piper command when the bundle is loaded
    to configure.

* == Used by default

========
Commands
========
/load <in/out> <bundle>: 
  Loads a input/output filter bundle named <bundle>.  If that bundle
  is already loaded, it will move it to the end of the list.

/unload <in/out> <bundle>:
  Unloads a input/output filter bundle named <bundle>

/loaded :
  Shows what bundles are currently loaded

/saveloaded :
  Saves the currently loaded bundles to the defaults so that
  the same ones will be loaded upon the next startup of 
  TalkSoup.app

/join <channel1>[,channel2,...] [password1[,password2,...]]
  Joins a channel (optionally 2 or more channels separated
  by commas with no space) with an optional password (or
  if there are multiple channels specified, then multiple
  passwords separated only by a comma)

/nick <nick>
  Changes your nickname to <nick>

/msg <nick> <message>
  Sends the private message <message> to the person named <nick>

/quit <message>
  Disconnects you from the currently connected IRC server.  If 
  <message> is specified, it will use that as the quit message.

/notice <nick> <message>
  Sends the private notice <message> to the person named <nick>
  Most of the time, you should send people regular messages using
  the /msg command.  However, in certain cases, you may want to
  use a notice instead.

/away <message>
  This command sets your away message if <message> is specified.  If you
  don't specify <message>, it will mark you as back.

/colors
  This simply explains to you the valid colors that other commands may
  ask you to use.

/ctcp <nick> <ctcp> [args]
  Sends the ctcp command <ctcp> with optional arguments [args] to
  the person with the nick <nick>.  CTCP messages are used for a
  lot of extensions to IRC such as DCC and getting information
  from another person's IRC client.

/version <nick>
  This is a sortcut for /ctcp <nick> VERSION, which, on most
  IRC clients, will return the person named <nick>'s IRC client 
  and version.

/clientinfo <nick>
  This is a shortcut for /ctcp <nick> CLIENTINFO, which, on
  most IRC clients, will return information on getting the
  IRC client that the person named <nick> is using.

/ping <nick> <argument>
  This is a shortcut for /ctcp <nick> PING <argument>.  With
  most IRC clients, the other client will respond immediately
  with a PING CTCP message with the same argument.  You can
  use this to see how lagged you are.

/userinfo <nick>
  This is a shortcut for /ctcp <nick> USERINFO.  On some
  IRC clients this will respond with information about the
  user named <nick>.

/topic <channel> [topic]
  This will set <channel>'s topic to [topic].  If [topic] is
  not specified, it will refresh the channel's topic.

/kick <channel> <user> [comment]
  This will kick the specified <user> out of a <channel>.  If
  [comment] is specified, it will be displayed to the user
  instead of the default message.

/raw <message>
  This command will send <message> to the server as it is, 
  no formatting or command substitution will take place on
  the string.

/encoding <encoding>
  This command changes the encoding for incoming and
  outgoing messages on that particular connection.  By not
  specifying <encoding> the command will list the available
  encodings.

==============================================================
Commands specific to the GNUstep output plugin (GNUstepOutput)
==============================================================

/me <action>
  Sends an action to the channel, for example:
  
    /me runs around the channel screaming 'GNUstep rules'
  
  would print * <nick> runs around the channel screaming 'GNUstep rules'

/query <nick>
  Opens a new tab for a person named <nick>.  All of their private 
  messages will appear in there, and typing to the person in that tab
  will send them private messages.

/server <server> [port]
  Disconnects from current server and connects to <server>.  If
  [port] is unspecified, it defaults to 6667.

/close [name]
  This closes the tab named [name].  If you don't specify the name, it will
  close the current tab.

/clear
  Clears the text in the current tab.

/scrollback <lines>
  This sets how many lines are in the scroll-back buffer.
  This change will automatically be saved to the defaults.

/ping <person>
  Good way to find out the lag between you and that person.

/topic [topic]
  This will set the current channel's topic to [topic].  If [topic] is
  not specified, the topic for the channel will be displayed.

/exec [-o] command
  This will execute command and display it's output.  If -o is specified,
  the output will actually be sent to the current channel/query.

/alias <alias> <command>
  Alias a command to another command.  For example,

  /alias j /join

  This will make /j the same as /join.

  /alias hi hello everybody!!!!

  This will make /hi send "hello everybody!!!!" to the channel/query.

/unalias <alias>
  Unset an alias set with the /alias command.
  
===================================
Notes about GNUstepOutput (default)
===================================
  
  * Double clicking a person in the user list in a channel will
    open a query tab for them.
  
  * See the 'Bundle Setup' to configure loaded bundles.  Some of the 
    bundles even provide nice GUIs for configuration.
	
===============================================================================
The FAQ (Frequently Asked Questions) also covers many common questions
that TalkSoup users have.
===============================================================================

=======
Authors
=======
  Main author: Andrew Ruder <aeruder at ksu dot edu>

==============
Special Thanks
==============
  Luis Garcia Alanis <luis.garcia at mac.com>
	For his initial drive to make a GNUstep IRC client using netclasses and his
	first version of Charla.app
  Alex Perez <aperez at student.santarosa.edu>
	For his immense testing efforts and all his suggestions and he also drew
	the icon that is currently being used by the GNUstep output plugin right
	now.  Alex is also one of the primary people behind making the original OS
	X port possible.
  Jeff Reinecke
	Jeff provided the machine that the port was done on.  Using solely VNC and
	SSH into his machine, the very first TalkSoup OS X was made possible.
	Thanks a ton Jeff!!!!
  Lyndon Tremblay <humasect at shaw.ca>
	For his time and effort in both testing and making TalkSoup an icon.
  Gregory Casemento <greg_casamento at yahoo.com>
	For his tremendous work on Gorm, which made TalkSoup possible.
  Alexander Malmberg <alexander at malmberg.org>
	For his constant help, bug-fixes, and overall contributions to the GNUstep
	project. 
  Matt Rice:
    He knows something about everything; a constant source of advice and help
	during the development of TalkSoup.
  Aaron Glenn:
    Actually made several of the new TalkSoup OS X nibs with me explaining
	(over IRC, of course) how to make them.  (He was unfamiliar with 
	using Interface Builder).  Without this guy, it's unlikely I would have
	ever gotten started on the new TalkSoup OS X port.
  
  Also, a big thanks goes out to: Nicolas Roard, Thom Cherryhomes, Tim
  Harrison, Ian Jones, and all the other people I probably forgot about who
  offered bug fixes, suggestions, and just took the time to compile and use
  TalkSoup.
    
===============================================================================
Copyright (C) 2003 by Andy Ruder
aeruder@ksu.edu
http://talksoup.aeruder.net
	


ChangeLog
2005-09-06 Andy Ruder <aeruder at ksu dot edu>
 * TalkSoup 1.0alpha 

2005-07-14 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput/BundlePreferences: Moved help text to external file,
 corrected help text.  Corrected NSTableView handling...

 * GNUstepOutput/Misc: Added code to debug NSView hierarchies

 * GNUstepOutput: Complete overhaul of NSTextView initializations, fixing
 of some corrupted gorms, making text field take on chat font, and other 
 bug fixes.
 
2005-07-12 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput
   NetclassesInput: Made changes to allow default quit message to work
   when quitting the app, closing a window, and /quit.

2005-07-07 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Moved to using distributed notifications instead of 
 distributed objects.
 * GNUstepOutput: Fix autoresize of nickview
 * GNUstepOutput: Fix tab completion picking up noncommand commandTyping:

2005-06-27 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: The NSTextField was much too much of a pain to use anymore.
 I've replaced the typing view full out with a KeyTextView which I think will
 also allow OS X to work sanely.

2005-06-26 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: On OS X focusing a textfield causes the action to be sent.
 This was causing the messages to be sent everytime the window was made key.
 Instead I am now trapping the key in the field editor so that it works 
 correctly on OS X and GNUstep.

2005-06-25 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Fixed serverlist focusing issues.  When you remove a server
 it won't always scroll to the top.

 * GNUstepOutput: Added 5 .nibs (special thanks to Aaron Glenn for doing this
 over IRC ;)

2005-06-24 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: There was a bug in the sorting code for the userlist.
 
 The place for typing will now retain it's position when tabs are changed/
 windows are taken from being key.

2005-06-22 Andrew Ruder <aeruder@ksu.edu>
 * IRCSwarm: Updating to allow time constraints (to adjust time between
 events) and correct case on nicknames.

 * GNUstepOutput: Using bold font correctly.  (It was only appropriately
 applied when the preferences were updated.)

 Removed caching (over complicated logic) from NSAttributedString.

 Several bug fixes in TabMasterController.m along the way to reenabling
 the select previous/next tab/close tab buttons on the menu. (which now
 work)

 * TalkSoup.app: I am a fricking retard.  I had doubleReleaseCheck on.
 Yes.  That is why it was slow.  I... am... a... retard.
 
 Retard
 Retard
 Retard

 TalkSoup is fast again.

2005-06-20 Andrew Ruder <aeruder@ksu.edu>
 * All: Additional header changes/code tweaks to allow clean compilation
 on OS X.  (Still no .nib files though... :-/)
 
 * All: Improvements on reducing autorelease usage.

 * GNUstepOutput: Force use of NSMessagePort even when the default is
 something else.

2005-06-17 Andrew Ruder <aeruder@ksu.edu>
 * Testing: Added a IRCSwarm testing tool that can replay preprocessed
 xchat logs (perl scripts can be easily written to process any type of log).
 It will connect to a server and more or less replay those logs.  Useful for
 debugging large amounts of messages.
 
2005-06-13 Andrew Ruder <aeruder@ksu.edu>
 * All: Filled in lots of headers that will be needed for
 OS X compilation.  Fixed relative paths in GNUmakefile.preambles
 in Output/GNUstepOutput and Input/NetclassesInput.

 * GNUstepOutput: /topic command works again, fixed topic inspector to
 stop double querying the topic.

2005-06-12 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: dns_helper returns the reverse lookup of the hostname
 as well.

2005-06-09 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Added a helper tool class to ease use of external
 tools + distributed objects.

 Moved /server dns look ups to a helper tool

 Added /alias,/unalias and a default set of aliases

 * NetclassesInput: Added a warning when using a package version < 1.04

 * Highlighting: Added back in the preferences to new preferences system.

 * DCCSupport: Added back in the preferences to new preferences system

2005-06-08 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Changed /scrollback lines to work with new scrollback
 history mechanism.  Obsoleted several unused preferences. ;)

 * GNUstepOutput: Added Tools/ directory for various helper tools.
 Began exec_helper (A tool to help with /exec)

 * GNUstepOutput: Added /exec and /exec -o commands

2005-06-05 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Fixed history corruption when closing tab with /close.
 Fixed notifications when removing tabs so that new selection is sent out
 as well.

2005-06-02 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Added a method to InputController to allow a command to
 be executed (programmatically) without making it into the history or messing
 up what the person is currently typing.

 Added support for // to escape normal command handling:
 //dev/hda is cool
 would send
 /dev/hda is cool
 to the channel.

 Fixed autoconnect commands for autoconnected servers.

 Fixed InputController's hold on ConnectionController when
 a ContentController is given a new ConnectionController

 Got rid of some used methods in ConnectionController pertaining to
 InputController

 Made InputController stop throwing things in the history.

 Added Favorites item to the server list that always stays on top.

 Added support for s// style tab completion

2005-06-01 Andrew Ruder <aeruder@ksu.edu>
 * GNUstepOutput: Topic inspector did not correctly update
 when someone else applied a new topic.  Fixed several memory leaks
 between interactions with ConnectionController and the ContentControllers.

 Should drop connections correctly when window is closed.  Added a default
 quit message preference and added to General preferences UI.  Added a dealloc
 routine for the StandardContentController class.  Fixed double initialization
 of the UI for QueryViewControllers and ChannelViewControllers.


Source/commands.m
/***************************************************************************
                               commands.m 
                          -------------------
    begin                : Mon Dec 22 07:34:32 CST 2003
    copyright            : (C) 2005 by Andrew Ruder
    email                : aeruder@ksu.edu
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#import "commands.h"

#import <Foundation/NSString.h>
#import <Foundation/NSInvocation.h>
#import <Foundation/NSAttributedString.h>
#import <Foundation/NSBundle.h>
#import <Foundation/NSArray.h>
#import <Foundation/NSNull.h>
#import <Foundation/NSUserDefaults.h>
#import <Foundation/NSEnumerator.h>
#import <Foundation/NSMapTable.h>

#define MARK [NSNull null]
#define NO_CONNECT S2AS(_(@"Connect to a server before using this command"))

@implementation TalkSoup (Commands)
- (NSAttributedString *)commandSaveLoaded: (NSString *)args 
   connection: (id)connection
{
	[self savePluginList];

	return S2AS(_(@"The loaded bundles will now load automagically on TalkSoup startup."));
}
- (NSAttributedString *)commandLoaded: (NSString *)args connection: (id)connection
{
	return BuildAttributedString(_(@"Currently loaded bundles:\n"),
	  MARK, IRCBold, IRCBoldValue, _(@"Output: "), activatedOutput, @"\n",
	  MARK, IRCBold, IRCBoldValue, _(@"Input: "), activatedInput, @"\n",
	  MARK, IRCBold, IRCBoldValue, _(@"Output Filters: "), [[self activatedOutFilters]
	    componentsJoinedByString: @", "], @"\n",
	  MARK, IRCBold, IRCBoldValue, _(@"Input Filters: "), [[self activatedInFilters]
	    componentsJoinedByString: @", "], nil);
}
- (NSAttributedString *)commandLoad: (NSString *)args connection: (id)connection
{
	id x = [args separateIntoNumberOfArguments: 3];
	id first, second;
	id key;
	id array = nil;
	BOOL isIn = NO;
	
	[self refreshPluginList];
	
	if ([x count] < 1)
	{
		return S2AS(_(@"Usage: /load <in/out>"));
	}
	
	first = [x objectAtIndex: 0];
	
	if ([first isEqualToString: @"in"])
	{
		array = [inNames allKeys];
		isIn = YES;
	}
	else if ([first isEqualToString: @"out"])
	{
		array = [outNames allKeys];
	}
	else
	{
		return S2AS(_(@"Usage: /load <in/out>"));
	}
	
	second = ([x count] > 1) ? [x objectAtIndex: 1] : nil;

	if (second && [array containsObject: second]) 
	{
		key = second;
	}
	else if (second)
	{
		NSEnumerator *iter;
		iter = [array objectEnumerator];
		while ((key = [iter nextObject])) 
		{
			if ([[key lowercaseString] isEqualToString: second])
				break;
		}
	}
	
	if (!second || !key)
	{
		return BuildAttributedString(
		  _(@"Usage: /load <in/out> <filter>"), @"\n",
		  MARK, IRCBold, IRCBoldValue, _(@"Possible filters: "), 
		  [array componentsJoinedByString: @", "], nil);
	}
	
	if (isIn)
	{
		[self activateInFilter: key];
	}
	else
	{
		[self activateOutFilter: key];
	}
	
	return BuildAttributedString(key, _(@" loaded"), nil);
}
- (NSAttributedString *)commandUnload: (NSString *)args connection: (id)connection
{
	id x = [args separateIntoNumberOfArguments: 3];
	id first, second;
	id key;
	id array = nil;
	BOOL isIn = NO;
	
	if ([x count] < 1)
	{
		return S2AS(_(@"Usage: /unload <in/out>"));
	}
	
	first = [x objectAtIndex: 0];
	
	if ([first isEqualToString: @"in"])
	{
		array = [self activatedInFilters];
		isIn = YES;
	}
	else if ([first isEqualToString: @"out"])
	{
		array = [self activatedOutFilters];
	}
	else
	{
		return S2AS(_(@"Usage: /unload <in/out>"));
	}

	second = ([x count] > 1) ? [x objectAtIndex: 1] : nil;
	
	if (second && [array containsObject: second]) 
	{
		key = second;
	}
	else if (second)
	{
		NSEnumerator *iter;
		iter = [array objectEnumerator];
		while ((key = [iter nextObject])) 
		{
			if ([[key lowercaseString] isEqualToString: second])
				break;
		}
	}
	
	if (!second || !key)
	{
		return BuildAttributedString(
		  _(@"Usage: /unload <in/out> <filter>"), @"\n", 
		  MARK, IRCBold, IRCBoldValue, _(@"Possible filters: "), 
		  [array componentsJoinedByString: @", "], nil);
	}
	
	if (isIn)
	{
		[self deactivateInFilter: key];
	}
	else
	{
		[self deactivateOutFilter: key];
	}
	
	return BuildAttributedString(key, _(@" unloaded"), nil);
}
- (NSAttributedString *)commandJoin: (NSString *)aString connection: connection
{
	NSArray *x = [aString separateIntoNumberOfArguments: 3];
	id pass;
	
	if (!connection) return NO_CONNECT;
	
	if ([x count] == 0)
	{
		return S2AS(_(@"Usage: /join <channel1[,channel2...]> [password1[,password2...]]"));
	}
	
	pass = ([x count] == 2) ? [x objectAtIndex: 1] : nil;
	
	[_TS_ joinChannel: S2AS([x objectAtIndex: 0]) withPassword: S2AS(pass) 
	  onConnection: connection
	  withNickname: S2AS([connection nick])
	  sender: output];
	  
	return nil;
}
- (NSAttributedString *)commandMsg: (NSString *)aString connection: connection
{
	NSArray *x = [aString separateIntoNumberOfArguments: 2];
	
	if (!connection) return NO_CONNECT;
	
	if ([x count] < 2)
	{
		return S2AS(_(@"Usage: /msg <person> <message>"));
	}
	
	[_TS_ sendMessage: S2AS([x objectAtIndex: 1]) to: 
	  S2AS([x objectAtIndex: 0])
	  onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];

	return nil;
}
- (NSAttributedString *)commandPart: (NSString *)args connection: connection
{
	id x = [args separateIntoNumberOfArguments: 2];
	id name, msg;
	
	if (!connection) return NO_CONNECT;
	
	msg = name = nil;
	
	if ([x count] >= 1)
	{
		name = [x objectAtIndex: 0];
	}
	if ([x count] >= 2)
	{
		msg = [x objectAtIndex: 1];
	}
	
	if (!name)
	{
		return S2AS(_(@"Usage: /part <channel> [message]"));
	}
	
	[_TS_ partChannel: S2AS(name) withMessage: S2AS(msg) 
	  onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandNotice: (NSString *)aString connection: connection
{
	NSArray *x = [aString separateIntoNumberOfArguments: 2];
	
	if (!connection) return NO_CONNECT;
	
	if ([x count] < 2)
	{
		return S2AS(_(@"Usage: /notice <person> <message>"));
	}
	
	[_TS_ sendNotice: S2AS([x objectAtIndex: 1]) to: 
	  S2AS([x objectAtIndex: 0])
	  onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];

	return nil;
}
- (NSAttributedString *)commandAway: (NSString *)aString connection: connection
{
	NSArray *x = [aString separateIntoNumberOfArguments: 1];
	id y = nil;
	
	if (!connection) return NO_CONNECT;
	
	if ([x count] > 0)
	{
		y = [x objectAtIndex: 0];
	}
	
	[_TS_ setAwayWithMessage: S2AS(y) onConnection: connection
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- commandNick: (NSString *)aString connection: connection
{
	NSArray *x = [aString separateIntoNumberOfArguments: 2];
	
	if (!connection) return NO_CONNECT;
	
	if ([x count] == 0)
	{
		return S2AS(_(@"Usage: /nick <newnick>"));
	}
	
	[_TS_ changeNick: S2AS([x objectAtIndex: 0]) onConnection: connection
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandQuit: (NSString *)aString connection: connection
{
	if (!connection) return NO_CONNECT;
	
	[_TS_ quitWithMessage: S2AS(aString) onConnection: connection
	  withNickname: S2AS([connection nick]) sender: output];
	
	return nil;
}
- (NSAttributedString *)commandColors: (NSString *)aString connection: connection
{
	return BuildAttributedString(
	 _(@"Valid color names include any color from the following list: "),
	 [PossibleUserColors() componentsJoinedByString: @", "], @"\n",
	 _(@"Also, a string is valid if it is of the form 'custom [red] [green] [blue]' "),
	  _(@"where [red], [green], [blue] are the red, green, and blue "),
	  _(@"components of the color on a scale of 0 to 1000."), nil);
}  		  
- (NSAttributedString *)commandCtcp: (NSString *)command connection: connection
{
	id array;
	id ctcp;
	id args;
	id who;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 3];
	
	if ([array count] < 2)
	{
		return S2AS(_(@"Usage: /ctcp <nick> <ctcp> [arguments]")); 
	}

	args = ([array count] == 3) ? [array objectAtIndex: 2] : nil;
	
	ctcp = [[array objectAtIndex: 1] uppercaseString];
	who = [array objectAtIndex: 0];

	[_TS_ sendCTCPRequest: S2AS(ctcp) withArgument: S2AS(args)
	  to: S2AS(who) onConnection: connection 
	  withNickname: S2AS([connection nick]) sender: output];
	
	return nil;
}	
- (NSAttributedString *)commandVersion: (NSString *)command connection: connection
{
	id array;
	id who;
	
	array = [command separateIntoNumberOfArguments: 2];

	if (!connection) return NO_CONNECT;
	
	if ([array count] == 0)
	{
		return S2AS(_(@"Usage: /version <nick>"));
	}

	who = [array objectAtIndex: 0];
	
	[_TS_ sendCTCPRequest: S2AS(@"VERSION") withArgument: nil
	  to: S2AS(who) onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandClientinfo: (NSString *)command connection: connection
{
	id array;
	id who;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 2];

	if ([array count] == 0)
	{
		return S2AS(_(@"Usage: /clientinfo <nick>"));
	}

	who = [array objectAtIndex: 0];
	
	[_TS_ sendCTCPRequest: S2AS(@"CLIENTINFO") withArgument: nil
	  to: S2AS(who) onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandUserinfo: (NSString *)command connection: connection
{
	id array;
	id who;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 2];

	if ([array count] == 0)
	{
		return S2AS(_(@"Usage: /userinfo <nick>"));
	}

	who = [array objectAtIndex: 0];
	
	[_TS_ sendCTCPRequest: S2AS(@"USERINFO") withArgument: nil
	  to: S2AS(who) onConnection: connection 
	  withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandPing: (NSString *)command connection: connection
{
	id array;
	id who;
	id arg = nil;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 2];

	if ([array count] <= 1)
	{
		return S2AS(_(@"Usage: /ping <nick> <argument>"));
	}

	who = [array objectAtIndex: 0];
	arg = [array objectAtIndex: 1];
	
	[_TS_ sendCTCPRequest: S2AS(@"PING") withArgument: S2AS(arg)
	  to: S2AS(who) onConnection: connection withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandTopic: (NSString *)command connection: connection
{
	id array;
	id who;
	id arg = nil;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 2];

	if ([array count] < 1)
	{
		return S2AS(_(@"Usage: /topic <channel> [topic]"));
	}

	who = [array objectAtIndex: 0];
	
	if ([array count] > 1)
	{
		arg = [array objectAtIndex: 1];
	}
	
	[_TS_ setTopicForChannel: S2AS(who) to: S2AS(arg)
	  onConnection: connection withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandKick: (NSString *)command connection: connection
{
	id array;
	id who;
	id arg = nil;
	id chan;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 3];

	if ([array count] <= 1)
	{
		return S2AS(_(@"Usage: /kick <channel> <user> [comment]"));
	}

	who = [array objectAtIndex: 1];
	chan = [array objectAtIndex: 0];
	if ([array count] > 2)
	{
		arg = [array objectAtIndex: 2];
	}
	
	[_TS_ kick: S2AS(who) offOf: S2AS(chan) for: S2AS(arg)
	  onConnection: connection withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandRaw: (NSString *)command connection: connection
{
	id array;
	id arg = nil;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 1];

	if ([array count] == 0)
	{
		return S2AS(_(@"Usage: /raw <message>"));
	}

	arg = [array objectAtIndex: 0];
	
	[_TS_ writeRawString: S2AS(arg)
	  onConnection: connection withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandMode: (NSString *)command connection: connection
{
	id array;
	id mode;
	id arg = AUTORELEASE([NSMutableArray new]);
	id obj;
	int ind, max;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: -1];

	max = [array count];
	
	if (max <= 1)
	{
		return S2AS(_(@"Usage: /mode <object> <mode(s)> [arguments]"));
	}

	mode = [array objectAtIndex: 1];
	obj = [array objectAtIndex: 0];
	
	for (ind = 2; ind < max; ind++)
	{
		[arg addObject: S2AS([array objectAtIndex: ind])];
	}
	
	[_TS_ setMode: S2AS(mode) on: S2AS(obj) withParams: arg
	  onConnection: connection withNickname: S2AS([connection nick])
	  sender: output];
	
	return nil;
}
- (NSAttributedString *)commandEncoding: (NSString *)command connection: connection
{
	id array;
	id arg = nil;
	NSStringEncoding enc = 0;
	
	if (!connection) return NO_CONNECT;
	
	array = [command separateIntoNumberOfArguments: 2];
	
	if ([array count] > 0)
	{
		arg = [array objectAtIndex: 0];
		arg = [arg lowercaseString];
	}
	
	if (arg) enc = [_TS_ encodingForIdentifier: arg];
	
	if (!enc)
	{
		NSMutableAttributedString *string;
		const NSStringEncoding *iter;
		string = AUTORELEASE([NSMutableAttributedString new]);

		for (iter = [_TS_ allEncodings]; *iter; iter++)
		{
			NSAttributedString *thisone;
			if ([string length] > 0) 
				[string appendAttributedString: S2AS(@", ")];
			thisone = 
			  BuildAttributedString(@"(", MARK, IRCBold, IRCBoldValue,
			  [_TS_ identifierForEncoding: *iter], @")", 
			  [_TS_ nameForEncoding: *iter], nil);
			[string appendAttributedString: thisone];
		}
		
		return BuildAttributedString(_(@"Usage: /encoding <encoding #>"), @"\n", 
		  MARK, IRCBold, IRCBoldValue, _(@"Available encodings: "), 
		  string, @"\n", _(@"Current encoding: "), @"(", MARK, IRCBold, IRCBoldValue,
		  [_TS_ identifierForEncoding: [connection encoding]], @")", 
		  [_TS_ nameForEncoding: [connection encoding]], nil);
	}
	
	[connection setEncoding: enc];
	
	return S2AS(_(@"Ok."));
}
- (void)setupCommandList
{
#define ADD_COMMAND(_sel, _name) { id invoc; \
	invoc = [NSInvocation invocationWithMethodSignature: \
	  [self methodSignatureForSelector: \
	  (_sel)]]; \
	[invoc retainArguments];\
	[invoc setSelector: (_sel)];\
	[invoc setTarget: self];\
	[self addCommand: (_name) withInvocation: invoc];}

	ADD_COMMAND(@selector(commandLoad:connection:), @"load");
	ADD_COMMAND(@selector(commandUnload:connection:), @"unload");
	ADD_COMMAND(@selector(commandLoaded:connection:), @"loaded");
	ADD_COMMAND(@selector(commandSaveLoaded:connection:), @"saveloaded");
	ADD_COMMAND(@selector(commandJoin:connection:), @"join");
	ADD_COMMAND(@selector(commandMsg:connection:), @"msg");
	ADD_COMMAND(@selector(commandPart:connection:), @"part");
	ADD_COMMAND(@selector(commandNotice:connection:), @"notice");
	ADD_COMMAND(@selector(commandAway:connection:), @"away");
	ADD_COMMAND(@selector(commandQuit:connection:), @"quit");
	ADD_COMMAND(@selector(commandColors:connection:), @"colors");
	ADD_COMMAND(@selector(commandCtcp:connection:), @"ctcp");
	ADD_COMMAND(@selector(commandVersion:connection:), @"version");
	ADD_COMMAND(@selector(commandClientinfo:connection:), @"clientinfo");
	ADD_COMMAND(@selector(commandUserinfo:connection:), @"userinfo");
	ADD_COMMAND(@selector(commandPing:connection:), @"ping");
	ADD_COMMAND(@selector(commandTopic:connection:), @"topic");
	ADD_COMMAND(@selector(commandKick:connection:), @"kick");
	ADD_COMMAND(@selector(commandRaw:connection:), @"raw");
	ADD_COMMAND(@selector(commandMode:connection:), @"mode");
	ADD_COMMAND(@selector(commandEncoding:connection:), @"encoding");
	
#undef ADD_COMMAND
}
@end

#undef MARK

